### PR TITLE
updated x509 certificates to x509v3 and SANs extension

### PR DESCRIPTION
**Pull request by alexbathome**
# Background

While implementing my own Go client for multipass, I noticed that the TLS certificates used by `multipassd`'s RPC server weren't using the SANs x509v3 extension. 

As such you would run into an annoying error when attempting to verify the TLS certificate that was being returned by the RPC server (`multipassd`). 

Go's standard library hasn't supported TLS/x509 certitficates without SANs present for a while now. See https://go.dev/doc/go1.16#minor_library_changes

# What does this change do? 

This change is a modification to the `ssl_cert_provider.cpp`'s x509 certificate generation. Here there are two major changes: 
1. Version of x509 certificate has been bumped to x509v3
2. Added SANs x509v3 Extension to include `DNS:${server_name}`

# Testing

I have managed to get this building and testing locally on my Ubuntu laptop, however I don't have any additional platforms available to me at the moment. 

The certificate is parsed successfully by `openssl x509 -in <>`  the output looks like:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: ${BLAH}
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: C = US, O = Canonical, CN = localhost
        Validity
            Not Before: Mar 31 10:42:15 2023 GMT
            Not After : Mar 30 10:42:15 2024 GMT
        Subject: C = US, O = Canonical, CN = localhost
        Subject Public Key Info:d
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                  ${BLAH}
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Subject Alternative Name: 
                DNS:localhost
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
      ${BLAH}
```

I have also verified the certificate loads correctly using a small GoLang program which simply establishes a connection against the multipassd unix socket.

# Additional notes

This is a duplicate/updated PR of #3004 as the original commits were not pushed with a correct author's email. 

If there's any questions, or concerns feel free to drop a comment in, I'm happy to facilitate any modifications to help this get merged :) 

Thanks! 

Fixes canonical/multipass#3005

### CODE REVIEW

**Comment by alexbathome** [comment]
@townsend2010 I have re-raised the PR here with the amended commits with the correct email (I hope). Are we able to check that the CLA check passes here? :pray: 

Many thanks

**Comment by townsend2010** [comment]
Hi @alexbathome,

I will run CI and if all goes well, we'll use this PR/branch and I'll close the other one.  Thanks!

**Comment by townsend2010** [comment]
Hi @alexbathome!

It's failing the code linting.  Please apply the following to fix t:
```diff
--- src/cert/ssl_cert_provider.cpp	(before formatting)
+++ src/cert/ssl_cert_provider.cpp	(after formatting)
@@ -118,12 +118,12 @@
     return server_name;
 }
 
-void set_san_name(X509 *c, std::string server_name)
+void set_san_name(X509* c, std::string server_name)
 {
     std::string san_dns = server_name;
-    GENERAL_NAMES *gens = sk_GENERAL_NAME_new_null();
-    GENERAL_NAME *gen = GENERAL_NAME_new();
-    ASN1_IA5STRING *ia5 = ASN1_IA5STRING_new();
+    GENERAL_NAMES* gens = sk_GENERAL_NAME_new_null();
+    GENERAL_NAME* gen = GENERAL_NAME_new();
+    ASN1_IA5STRING* ia5 = ASN1_IA5STRING_new();
     ASN1_STRING_set(ia5, san_dns.data(), san_dns.length());
     GENERAL_NAME_set0_value(gen, GEN_DNS, ia5);
     sk_GENERAL_NAME_push(gens, gen);
@@ -165,7 +165,6 @@
         X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, cn.data(), cn.size(), APPEND_ENTRY, ADD_RDN);
         X509_set_issuer_name(x509.get(), name);
         set_san_name(x509.get(), server_name);
-        
 
         if (!X509_set_pubkey(x509.get(), key.get()))
             throw std::runtime_error("Failed to set certificate public key");
```

**Comment by townsend2010** [comment]
BTW, the CLA check is all good now!  Thanks!

**Comment by alexbathome** [comment]
thanks @townsend2010 you're formatting patch there has been applied now; let's see how CI goes now :pray:

File: src/cert/ssl_cert_provider.cpp
Comment (townsend2010) [inline src/cert/ssl_cert_provider.cpp:121]:
```suggestion
void set_san_name(X509* c, const std::string& server_name)
```
We prefer const reference for this.

**Review by townsend2010 — CHANGES_REQUESTED** [review]
Hi @alexbathome!

Looks good overall and works well, thank you!  I have one small niggle inline and then after that, I'll get this merged.

Thanks again!

File: src/cert/ssl_cert_provider.cpp
Comment (alexbathome) [reply src/cert/ssl_cert_provider.cpp]: Resolved in latest commit :pray: :rocket:

File: src/cert/ssl_cert_provider.cpp
Comment (townsend2010) [inline src/cert/ssl_cert_provider.cpp:156]: Oh, I missed this.  Please leave in the white space to better differentiate the methods.

**Review by townsend2010 — CHANGES_REQUESTED** [review]
Hi @alexbathome!

One more little thing below.  Sorry for all of the back and forth, but we are super close on this!  Thanks for your persistence!

**Comment by alexbathome** [comment]
> Uh oh, I think there was a bad rebase and push? I see many more commits now.

Apologies for that; should be all resolved now for you there Chris!

Thanks! :sweat:

**Review by townsend2010 — APPROVED** [review]
Ok, LGTM!  I'll wait for the CI run to go through it's paces and then I'll merge it once that is good.  Thanks again!

**Comment by townsend2010** [comment]
bors merge

**Comment by townsend2010** [comment]
Oh, right, bors doesn't run on contributor branches.  Will merge manually.  Thanks again!

**Pull request merged**
